In [2]:
import json
import os
import pandas as pd
import numpy as np
from datetime import datetime 
import matplotlib.pyplot as plt
import re
from matplotlib.pyplot import figure

In [8]:
output_dir = "error_processed_data/"
file = "error_tree.pki"
raw_data = pd.read_pickle(output_dir + file)
raw_data = raw_data[["time", "User Name", "command", "full_command", "stderr"]]
raw_data

time User Name command  \
0   2021-04-19 19:34:00    User_1      ls   
1   2021-04-19 19:35:00    User_1  export   
2   2021-04-19 19:35:00    User_1  export   
3   2021-04-19 19:36:00    User_1  export   
4   2021-04-19 19:36:00    User_1  export   
..                  ...       ...     ...   
341 2023-01-19 14:40:00      Lily   touch   
342 2022-12-15 19:11:00     Megan   clear   
343 2022-12-15 19:11:00     Megan      ls   
344 2022-12-15 19:14:00     Megan      cd   
345 2022-12-15 19:16:00     Megan   clear   

                                          full_command  \
0                                                   ls   
1                                   export $TEST=Hello   
2                                 export $TEST='Hello'   
3                          export PATH=$PATH:$HOME/bin   
4                           export PATH=$PATH:test/bin   
..                                                 ...   
341  touch ~/data-store/data/input/Discovery/TSVFil...   
342                                              clear   
343                                                 ls   
344                                                 cd   
345                                              clear   

                                                stderr  
0                                                       
1    /home/supershell/.offline.sh: line 358: export...  
2    /home/supershell/.offline.sh: line 358: export...  
3                                                       
4                                                       
..                                                 ...  
341  touch: cannot touch '/home/jovyan/data-store/d...  
342                                                     
343                                                     
344                                                     
345                                                     

[819 rows x 5 columns]

In [9]:
def error_splitter(errors):
    return errors.splitlines()
raw_data["error list"] = raw_data["stderr"].apply(lambda x: error_splitter(x))
raw_data

time User Name command  \
0   2021-04-19 19:34:00    User_1      ls   
1   2021-04-19 19:35:00    User_1  export   
2   2021-04-19 19:35:00    User_1  export   
3   2021-04-19 19:36:00    User_1  export   
4   2021-04-19 19:36:00    User_1  export   
..                  ...       ...     ...   
341 2023-01-19 14:40:00      Lily   touch   
342 2022-12-15 19:11:00     Megan   clear   
343 2022-12-15 19:11:00     Megan      ls   
344 2022-12-15 19:14:00     Megan      cd   
345 2022-12-15 19:16:00     Megan   clear   

                                          full_command  \
0                                                   ls   
1                                   export $TEST=Hello   
2                                 export $TEST='Hello'   
3                          export PATH=$PATH:$HOME/bin   
4                           export PATH=$PATH:test/bin   
..                                                 ...   
341  touch ~/data-store/data/input/Discovery/TSVFil...   
342                                              clear   
343                                                 ls   
344                                                 cd   
345                                              clear   

                                                stderr  \
0                                                        
1    /home/supershell/.offline.sh: line 358: export...   
2    /home/supershell/.offline.sh: line 358: export...   
3                                                        
4                                                        
..                                                 ...   
341  touch: cannot touch '/home/jovyan/data-store/d...   
342                                                      
343                                                      
344                                                      
345                                                      

                                            error list  
0                                                   []  
1    [/home/supershell/.offline.sh: line 358: expor...  
2    [/home/supershell/.offline.sh: line 358: expor...  
3                                                   []  
4                                                   []  
..                                                 ...  
341  [touch: cannot touch '/home/jovyan/data-store/...  
342                                                 []  
343                                                 []  
344                                                 []  
345                                                 []  

[819 rows x 6 columns]

In [10]:
expand_data = raw_data.explode('error list')
expand_data

time User Name command  \
0   2021-04-19 19:34:00    User_1      ls   
1   2021-04-19 19:35:00    User_1  export   
2   2021-04-19 19:35:00    User_1  export   
3   2021-04-19 19:36:00    User_1  export   
4   2021-04-19 19:36:00    User_1  export   
..                  ...       ...     ...   
341 2023-01-19 14:40:00      Lily   touch   
342 2022-12-15 19:11:00     Megan   clear   
343 2022-12-15 19:11:00     Megan      ls   
344 2022-12-15 19:14:00     Megan      cd   
345 2022-12-15 19:16:00     Megan   clear   

                                          full_command  \
0                                                   ls   
1                                   export $TEST=Hello   
2                                 export $TEST='Hello'   
3                          export PATH=$PATH:$HOME/bin   
4                           export PATH=$PATH:test/bin   
..                                                 ...   
341  touch ~/data-store/data/input/Discovery/TSVFil...   
342                                              clear   
343                                                 ls   
344                                                 cd   
345                                              clear   

                                                stderr  \
0                                                        
1    /home/supershell/.offline.sh: line 358: export...   
2    /home/supershell/.offline.sh: line 358: export...   
3                                                        
4                                                        
..                                                 ...   
341  touch: cannot touch '/home/jovyan/data-store/d...   
342                                                      
343                                                      
344                                                      
345                                                      

                                            error list  
0                                                  NaN  
1    /home/supershell/.offline.sh: line 358: export...  
2    /home/supershell/.offline.sh: line 358: export...  
3                                                  NaN  
4                                                  NaN  
..                                                 ...  
341  touch: cannot touch '/home/jovyan/data-store/d...  
342                                                NaN  
343                                                NaN  
344                                                NaN  
345                                                NaN  

[830 rows x 6 columns]

**Analyze on preceding commands of error**

In [11]:
file_name_replacer = re.compile(r'.+(\').+(\').+')
file_name = re.compile(r'(\').+(\')')

In [12]:
path_replacer = re.compile(r'.*(\/).+(: line )(\d+).*')
path_name = re.compile(r'(\/).+(: line )(\d+)')

In [13]:
name_between_colon = re.compile(r'.+(:).+(:).+')
colon_name = re.compile(r'(:).+(:)')

In [14]:
cd_replacer = re.compile(r'.*cd:(.+):.*')
cd_name = re.compile(r'cd:(.+):')

In [15]:
def replacer(error):
    if file_name_replacer.match(error):
        return re.sub(file_name, '$', error)
    elif path_replacer.match(error):
        error = re.sub(path_name, 'PATH', error)
        if file_name_replacer.match(error):
            return re.sub(file_name, '$', error)
        elif cd_replacer.match(error):
            return re.sub(cd_name, 'cd: $:', error)
        else:
            return error
    elif name_between_colon.match(error):
        return re.sub(colon_name, ': $:', error)

***Choose cat: cannot access $: No such file or directory as an example***

In [17]:
error_filter = (error_expand["no name"] == "cat: $: No such file or directory")

In [18]:
cat_error = error_expand[error_filter]
cat_error.head()

time User Name                   command  \
68  2021-04-16 23:13:00    User_9                       cat   
97  2021-04-17 13:50:00    User_9                       cat   
172 2021-04-18 22:05:00    User_4  directory_grep_head_tail   
176 2021-04-18 22:05:00    User_4  directory_grep_head_tail   
410 2022-09-11 20:13:00    User_3                       cat   

                  full_command  \
68    cat "!!!!" > goodbye.txt   
97             cat goodbye.txt   
172                        NaN   
176                        NaN   
410  cat hello.txt goodbye.txt   

                                                stderr  \
68              cat: '!!!!': No such file or directory   
97         cat: goodbye.txt: No such file or directory   
172  cat: /home/jovyan/Bash/FilteredSortedOutput/Se...   
176  cat: /home/jovyan/Bash/FilteredSortedOutput/Se...   
410        cat: goodbye.txt: No such file or directory   

                                            error list  \
68            [cat: '!!!!': No such file or directory]   
97       [cat: goodbye.txt: No such file or directory]   
172  [/home/jovyan/Bash/bin/directory_grep_head_tai...   
176  [/home/jovyan/Bash/bin/directory_grep_head_tai...   
410      [cat: goodbye.txt: No such file or directory]   

                               no name  
68   cat: $: No such file or directory  
97   cat: $: No such file or directory  
172  cat: $: No such file or directory  
176  cat: $: No such file or directory  
410  cat: $: No such file or directory

**Cat 1: cat "!!!">goodbye.txt**

In [19]:
cat_error_68 = error_expand.iloc[65:72]
cat_error_68

time User Name command                         full_command  \
65 2021-04-16 23:13:00    User_9     cat                      cat goodbye.txt   
66 2021-04-16 23:13:00    User_9     cat                 cat link_goodbye.txt   
67 2021-04-16 23:13:00    User_9    echo           echo "!!!!" >> goodbye.txt   
68 2021-04-16 23:13:00    User_9     cat             cat "!!!!" > goodbye.txt   
69 2021-04-16 23:15:00    User_9     cat                 cat link_goodbye.txt   
70 2021-04-16 23:15:00    User_9    pico                                  NaN   
71 2021-04-16 23:15:00    User_9    diff  diff goodbye.txt copied_goodbye.txt   

                                    stderr  \
65                                           
66                                           
67                                           
68  cat: '!!!!': No such file or directory   
69                                           
70                                           
71                                           

                                  error list  \
65                                        []   
66                                        []   
67                                        []   
68  [cat: '!!!!': No such file or directory]   
69                                        []   
70                                        []   
71                                        []   

                              no name  
65                               None  
66                               None  
67                               None  
68  cat: $: No such file or directory  
69                               None  
70                               None  
71                               None

**Cat 2: cat goodbye.txt**

In [20]:
cat_error_97 = error_expand.iloc[94:101]
cat_error_97

time User Name command                        full_command  \
94  2021-04-17 13:50:00    User_9      ls                               ls -l   
95  2021-04-17 13:50:00    User_9      mv  mv goodbye.txt renamed_goodbye.txt   
96  2021-04-17 13:50:00    User_9      ls                               ls -l   
97  2021-04-17 13:50:00    User_9     cat                     cat goodbye.txt   
98  2021-04-17 13:57:00    User_9      ls                               ls -r   
99  2021-04-17 13:57:00    User_9      ls                            ls -l -r   
100 2021-04-17 13:57:00    User_9      ls                               ls -R   

                                          stderr  \
94                                                 
95                                                 
96                                                 
97   cat: goodbye.txt: No such file or directory   
98                                                 
99                                                 
100                                                

                                        error list  \
94                                              []   
95                                              []   
96                                              []   
97   [cat: goodbye.txt: No such file or directory]   
98                                              []   
99                                              []   
100                                             []   

                               no name  
94                                None  
95                                None  
96                                None  
97   cat: $: No such file or directory  
98                                None  
99                                None  
100                               None

Error results because type the wrong file name

**Cat 3: cat hello.txt**

In [21]:
cat_error_410 = error_expand.iloc[407:414]
cat_error_410

time User Name command               full_command  \
407 2022-09-11 20:12:00    User_3     pwd                        pwd   
408 2022-09-11 20:13:00    User_3     cat   cat hello.txt goobye.txt   
409 2022-09-11 20:13:00    User_3      ls                         ls   
410 2022-09-11 20:13:00    User_3     cat  cat hello.txt goodbye.txt   
411 2022-09-11 20:13:00    User_3     cat              cat hello.txt   
412 2022-09-11 20:20:00    User_3     cat                        cat   
413 2022-09-11 20:22:00    User_3      ls                ls > ls.out   

                                          stderr  \
407                                                
408                                                
409                                                
410  cat: goodbye.txt: No such file or directory   
411                                                
412                                                
413                                                

                                        error list  \
407                                             []   
408                                             []   
409                                             []   
410  [cat: goodbye.txt: No such file or directory]   
411                                             []   
412                                             []   
413                                             []   

                               no name  
407                               None  
408                               None  
409                               None  
410  cat: $: No such file or directory  
411                               None  
412                               None  
413                               None

Error results because type the wrong file name. a Typo

**Another Example: ls: cannot access $: No such file or directory**

In [22]:
error_filter_ls = (error_expand["no name"] == "ls: cannot access $: No such file or directory")

In [23]:
ls_error = error_expand[error_filter_ls]
ls_error.head(10)

time User Name command                     full_command  \
58  2021-04-16 22:56:00    User_9      ls    ls -l misch_child/goodbye.txt   
264 2022-09-11 18:27:00    User_3      ls  ls /home/jovyan/RNA_SEQ_SAMPLES   
270 2022-09-11 18:28:00    User_3      ls      ls /gunings/RNA_SEQ_SAMPLES   
285 2022-09-11 18:33:00    User_3      ls                   ls stdout,txt.   
286 2022-09-11 18:35:00    User_3      ls                          ls RNAt   
287 2022-09-11 18:37:00    User_3      ls                           ls RNA   
288 2022-09-11 18:37:00    User_3      ls                         ls RNA^t   
290 2022-09-11 18:51:00    User_3      ls                         ls RNA^t   
299 2022-09-11 18:57:00    User_3      ls                            ls –l   
302 2022-09-11 19:17:00    User_3      ls                 ls –l stdout.txt   

                                                stderr  \
58   ls: cannot access 'misch_child/goodbye.txt': N...   
264  ls: cannot access '/home/jovyan/RNA_SEQ_SAMPLE...   
270  ls: cannot access '/gunings/RNA_SEQ_SAMPLES': ...   
285  ls: cannot access 'stdout,txt.': No such file ...   
286  ls: cannot access 'RNAt': No such file or dire...   
287  ls: cannot access 'RNA': No such file or direc...   
288  ls: cannot access 'RNA^t': No such file or dir...   
290  ls: cannot access 'RNA^t': No such file or dir...   
299  ls: cannot access '–l': No such file or directory   
302  ls: cannot access '–l': No such file or directory   

                                            error list  \
58   [ls: cannot access 'misch_child/goodbye.txt': ...   
264  [ls: cannot access '/home/jovyan/RNA_SEQ_SAMPL...   
270  [ls: cannot access '/gunings/RNA_SEQ_SAMPLES':...   
285  [ls: cannot access 'stdout,txt.': No such file...   
286  [ls: cannot access 'RNAt': No such file or dir...   
287  [ls: cannot access 'RNA': No such file or dire...   
288  [ls: cannot access 'RNA^t': No such file or di...   
290  [ls: cannot access 'RNA^t': No such file or di...   
299  [ls: cannot access '–l': No such file or direc...   
302  [ls: cannot access '–l': No such file or direc...   

                                            no name  
58   ls: cannot access $: No such file or directory  
264  ls: cannot access $: No such file or directory  
270  ls: cannot access $: No such file or directory  
285  ls: cannot access $: No such file or directory  
286  ls: cannot access $: No such file or directory  
287  ls: cannot access $: No such file or directory  
288  ls: cannot access $: No such file or directory  
290  ls: cannot access $: No such file or directory  
299  ls: cannot access $: No such file or directory  
302  ls: cannot access $: No such file or directory

**ls example: 58**

In [24]:
ls_error_58 = error_expand.iloc[55:62]
ls_error_58

time User Name command                       full_command  \
55 2021-04-16 22:56:00    User_9      cd                              cd ..   
56 2021-04-16 22:56:00    User_9      ls                                 ls   
57 2021-04-16 22:56:00    User_9      ln       ln -s goodbye.txt misc_child   
58 2021-04-16 22:56:00    User_9      ls      ls -l misch_child/goodbye.txt   
59 2021-04-16 22:56:00    User_9      ls                                 ls   
60 2021-04-16 22:57:00    User_9      ls                   ls -l misc_child   
61 2021-04-16 23:05:00    User_9      cp  cp goodbye.txt copied_goodbye.txt   

                                               stderr  \
55                                                      
56                                                      
57                                                      
58  ls: cannot access 'misch_child/goodbye.txt': N...   
59                                                      
60                                                      
61                                                      

                                           error list  \
55                                                 []   
56                                                 []   
57                                                 []   
58  [ls: cannot access 'misch_child/goodbye.txt': ...   
59                                                 []   
60                                                 []   
61                                                 []   

                                           no name  
55                                            None  
56                                            None  
57                                            None  
58  ls: cannot access $: No such file or directory  
59                                            None  
60                                            None  
61                                            None

Reason of error: type wrong path name

**ls example 264**

In [26]:
ls_error_264 = error_expand.iloc[261:270]
ls_error_264

time User Name command                     full_command  \
261 2022-03-02 20:57:00    User_8      ls                 ls /home/jovyan/   
262 2022-03-02 20:58:00    User_8      ls             ls /home/jovyan/data   
263 2022-09-11 18:18:00    User_3      ls                  ls /home/jovyan   
264 2022-09-11 18:27:00    User_3      ls  ls /home/jovyan/RNA_SEQ_SAMPLES   
265 2022-09-11 18:27:00    User_3      ls                            ls ..   
266 2022-09-11 18:27:00    User_3      ls                               ls   
267 2022-09-11 18:28:00    User_3      ls                               ls   
268 2022-09-11 18:28:00    User_3      cd                          cd home   
269 2022-09-11 18:28:00    User_3      ls                               ls   

                                                stderr  \
261                                                      
262                                                      
263                                                      
264  ls: cannot access '/home/jovyan/RNA_SEQ_SAMPLE...   
265                                                      
266                                                      
267                                                      
268                                                      
269                                                      

                                            error list  \
261                                                 []   
262                                                 []   
263                                                 []   
264  [ls: cannot access '/home/jovyan/RNA_SEQ_SAMPL...   
265                                                 []   
266                                                 []   
267                                                 []   
268                                                 []   
269                                                 []   

                                            no name  
261                                            None  
262                                            None  
263                                            None  
264  ls: cannot access $: No such file or directory  
265                                            None  
266                                            None  
267                                            None  
268                                            None  
269                                            None

Reason of error: wrong path

**ls example 285**

In [27]:
ls_error_285 = error_expand.iloc[282:290]
ls_error_285

time User Name command    full_command  \
282 2022-09-11 18:32:00    User_3      cd           cd ..   
283 2022-09-11 18:32:00    User_3      ls              ls   
284 2022-09-11 18:33:00    User_3      ls   ls stdout.txt   
285 2022-09-11 18:33:00    User_3      ls  ls stdout,txt.   
286 2022-09-11 18:35:00    User_3      ls         ls RNAt   
287 2022-09-11 18:37:00    User_3      ls          ls RNA   
288 2022-09-11 18:37:00    User_3      ls        ls RNA^t   
289 2022-09-11 18:51:00    User_3      cd         cd home   

                                                stderr  \
282                                                      
283                                                      
284                                                      
285  ls: cannot access 'stdout,txt.': No such file ...   
286  ls: cannot access 'RNAt': No such file or dire...   
287  ls: cannot access 'RNA': No such file or direc...   
288  ls: cannot access 'RNA^t': No such file or dir...   
289                                                      

                                            error list  \
282                                                 []   
283                                                 []   
284                                                 []   
285  [ls: cannot access 'stdout,txt.': No such file...   
286  [ls: cannot access 'RNAt': No such file or dir...   
287  [ls: cannot access 'RNA': No such file or dire...   
288  [ls: cannot access 'RNA^t': No such file or di...   
289                                                 []   

                                            no name  
282                                            None  
283                                            None  
284                                            None  
285  ls: cannot access $: No such file or directory  
286  ls: cannot access $: No such file or directory  
287  ls: cannot access $: No such file or directory  
288  ls: cannot access $: No such file or directory  
289                                            None

Reason of error: typo of file name. Should be period but use comma

**ls example 302**

In [28]:
ls_error_302 = error_expand.iloc[299:306]
ls_error_302

time User Name command      full_command  \
299 2022-09-11 18:57:00    User_3      ls             ls –l   
300 2022-09-11 18:57:00    User_3      ls             ls -l   
301 2022-09-11 19:17:00    User_3      ls  ls -l stdout.txt   
302 2022-09-11 19:17:00    User_3      ls  ls –l stdout.txt   
303 2022-09-11 19:17:00    User_3      ls                ls   
304 2022-09-11 19:17:00    User_3      cd             cd ..   
305 2022-09-11 19:17:00    User_3      ls                ls   

                                                stderr  \
299  ls: cannot access '–l': No such file or directory   
300                                                      
301                                                      
302  ls: cannot access '–l': No such file or directory   
303                                                      
304                                                      
305                                                      

                                            error list  \
299  [ls: cannot access '–l': No such file or direc...   
300                                                 []   
301                                                 []   
302  [ls: cannot access '–l': No such file or direc...   
303                                                 []   
304                                                 []   
305                                                 []   

                                            no name  
299  ls: cannot access $: No such file or directory  
300                                            None  
301                                            None  
302  ls: cannot access $: No such file or directory  
303                                            None  
304                                            None  
305                                            None

Error reason: cannot link